src/dmd/mars.d

```diff
@@ -1235,6 +1242,9 @@ void addDefaultVersionIdentifiers(const ref Param params)
     VersionCondition.addPredefinedGlobalIdent("D_Version2");
     VersionCondition.addPredefinedGlobalIdent("all");
 
+    if (params.cpp98)
+        VersionCondition.addPredefinedGlobalIdent("Cpp98");
```

Comment:
I don't think a version identifier should be set. I think it's more suitable to use `__traits(getTargetInfo)` for this.

Reply:
I think this is a good point. versions are boolean; a version token is either on or off, whereas this is a case where the C++ compatibility mode is a value from an enum...

Reply:
I can't see preferring:
```
static if (__traits(getTargetInfo, CppLevel) == 98)
```
over:
```
version (Cpp98)
```

Reply:
I think this is probably 90% use case:
```d
static if (__traits(getTargetInfo, CppStd) >= CppStd.Cpp11)
```
This thing is definitely an enum, not a set of bool's.

Reply:
@WalterBright If you want to use a `version`, then what do you do when you'll want to add `Cpp11` or `Cpp17`?

- Do you make `Cpp17` not imply `Cpp98`? Then, D code needs to be updated to check for both.

- Do you make `Cpp17` also automatically enable `Cpp98`? Then, what do you do in D code to detect `Cpp98` specifically? `version(Cpp98) { version(Cpp17) {} else { /*do C++98 thing*/ }}` is a mouthful, but also, it breaks if you decide to introduce `Cpp11` or another interim version later.

Manu is right - an enum solves both of these problems. Version numbers are points on a range, not a set of flags.

Reply:
```
version (Cpp98) {
    ...
} else version (Cpp17) {
   ...
} else {
   ... default ...
}
```
It'll be fine.

Reply:
Hardly, this will also make it more difficult to automatically port C++ code that uses `#if __cplusplus > whatever`.

Reply:
That doesn't make sense... There's 2 moments with respect to everything that might have version distinction; before X, and after X. Versions dont express this.

What is your motive to resist on this? I don't understand the criticism.

Reply:
> It'll be fine.

That doesn't tackle the problem, though. Do you duplicate the code in `Cpp17`? And what do you do when `Cpp11` is introduced?

> What is your motive to resist on this? I don't understand the criticism.

I'd be willing to guess 1) the implementation is much simpler 2) the implementation is already written.

Reply:
> That doesn't tackle the problem, though. Do you duplicate the code in Cpp17? And what do you do when Cpp11 is introduced?

The same thing we do with operating system `version`s. There's a very good reason why versions are binary, not a level. Levels result in bugs, because they are open-ended rather than closed. With that and the code duplication argument, I've gone over that in the forum innumerable times, though I can't seem to get the druntime code to follow the idea correctly. People can't let go of the brittle C way of doing it.

Lastly, the `Cpp98` is purposely looking backwards, not forwards. C++98 is not going to be changing, by definition. `Cpp98` means "be compatible with the old version which will not ever change", not "be compatible with the latest version whatever that may be". This is why the use is:
```
version (Cpp98)
   ... do it the old C++ 98 way ...
else
   ... do it the latest way ...
```
not:
```
version (Cpp11orNewer)
    ... do it the latest way ...
else
    ... do it the C++98 way ...
```
It's a different way of framing the problem, and in my not-so-humble opinion, a better way.

Reply:
I think it's a waste having to add a new version identifier for each version of C++. It's also receivers yet another version identifier that cannot be used by the user. BTW, therefore this is a breaking change. I might already be using the `Cpp98` version identifier and now my code will stop compiling since suddenly `Cpp98` is reserved. Using a trait prevents that.